File: app/src/test/load/RecommendationLoadScenario.java
package load;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import io.gatling.javaapi.core.ActionBuilder;
import io.gatling.javaapi.core.ChainBuilder;
import io.gatling.javaapi.core.CoreDsl;
import io.gatling.javaapi.core.PopulationBuilder;
import io.gatling.javaapi.core.ScenarioBuilder;
import io.gatling.javaapi.core.Simulation;
import io.gatling.javaapi.http.HttpDsl;
import io.gatling.javaapi.http.HttpProtocolBuilder;

import java.io.IOException;
import java.io.InputStream;
import java.time.Duration;
import java.util.UUID;

import static io.gatling.javaapi.core.CoreDsl.*;
import static io.gatling.javaapi.http.HttpDsl.*;

/**
 * Gatling load scenario for the AI Smart Dispatch recommendation endpoint.
 *
 * <h3>Purpose</h3>
 * Validates the Phase 2 exit criterion: p95 recommendation latency ≤ 3 seconds and
 * server error rate &lt; 1%, against a seeded pool of 200 technicians at twice the design
 * peak write rate.
 *
 * <h3>Profile selection</h3>
 * Set the JVM system property {@code load.profile} to override the profile:
 * <pre>
 *   mvn gatling:test -Dload.profile=degraded
 *   mvn gatling:test -Dload.profile=ci-smoke
 * </pre>
 * Defaults to {@code healthy} (the Phase 2 gate profile).
 *
 * <h3>Prerequisites</h3>
 * <ol>
 *   <li>The application must be running at the URL specified by {@code load.baseUrl}
 *       (default {@code http://localhost:8080}).</li>
 *   <li>The travel-time provider stub must be running at {@code load.travelStubUrl}
 *       (default {@code http://localhost:8099}) — see
 *       {@code src/test/resources/stubs/travel-matrix/} for the WireMock stub files.</li>
 *   <li>The database must have been seeded with
 *       {@code fixtures/seed-core.sql} and
 *       {@code fixtures/seed-load-200-technicians.sql}.</li>
 * </ol>
 *
 * <h3>Scenario mix (healthy profile, twice design peak)</h3>
 * <ul>
 *   <li>50% — GET /api/v1/work-orders/{id}/recommendations</li>
 *   <li>20% — POST /api/v1/work-orders (create)</li>
 *   <li>15% — GET /api/v1/work-orders (paginated search)</li>
 *   <li>10% — PATCH /api/v1/work-orders/{id}/transitions (state transition)</li>
 *   <li>5%  — POST /api/v1/work-orders/{id}/assignments</li>
 * </ul>
 *
 * <h3>Phases</h3>
 * All durations are read from {@code src/test/load/config/profiles.json}.
 * The warm-up phase is excluded from percentile calculation by Gatling's
 * {@code RampConcurrentUsers} + {@code atOnceUsers} warm-up pattern.
 *
 * <h3>Result file</h3>
 * After the simulation, a machine-readable {@code results.json} is written to
 * {@code target/load-results/}. Run {@link com.fieldservice.load.ThresholdEvaluator}
 * against it with {@code src/test/load/config/thresholds.json} to gate the build.
 *
 * @see com.fieldservice.load.ThresholdEvaluator
 */
public class RecommendationLoadScenario extends Simulation {

    private static final ObjectMapper MAPPER       = new ObjectMapper();
    private static final String       BASE_URL     = System.getProperty("load.baseUrl",       "http://localhost:8080");
    private static final String       PROFILE_NAME = System.getProperty("load.profile",       "healthy");
    private static final String       DISPATCHER_TOKEN  = System.getProperty("load.dispatcherToken", "test-dispatcher-token");

    // ── Profile loading ───────────────────────────────────────────────────────

    private static final JsonNode PROFILE = loadProfile();

    private static JsonNode loadProfile() {
        try (InputStream is = RecommendationLoadScenario.class
                .getResourceAsStream("/load/config/profiles.json")) {
            if (is == null) {
                // Fall back to embedded defaults if the resource is absent at classpath
                return MAPPER.readTree("""
                    {"profiles":{"healthy":{"warmUpSeconds":60,"rampUpSeconds":30,"holdSeconds":120,
                    "coolDownSeconds":15,"targetVirtualUsers":40}}}""");
            }
            return MAPPER.readTree(is).path("profiles").path(PROFILE_NAME);
        } catch (IOException e) {
            throw new IllegalStateException("Failed to load load profile '" + PROFILE_NAME + "'", e);
        }
    }

    private int profileInt(String key, int defaultValue) {
        JsonNode node = PROFILE.path(key);
        return node.isMissingNode() ? defaultValue : node.intValue();
    }

    // ── HTTP protocol ─────────────────────────────────────────────────────────

    private final HttpProtocolBuilder httpProtocol = http
            .baseUrl(BASE_URL)
            .acceptHeader("application/json")
            .contentTypeHeader("application/json")
            .header("Authorization", "Bearer " + DISPATCHER_TOKEN)
            .shareConnections()
            .disableCaching();

    // ── Work order IDs shared across scenario mix ──────────────────────────────
    // These are seeded by seed-load-200-technicians.sql companion WO seed.

    private static final String WO_NEW_1 = "cc000000-0000-7206-0020-000000000001";
    private static final String WO_NEW_2 = "cc000000-0000-7206-0020-000000000002";

    // ── Scenario builders ─────────────────────────────────────────────────────

    private final ChainBuilder getRecommendations =
            exec(http("GET recommendations (pool=200)")
                    .get("/api/v1/work-orders/" + WO_NEW_1 + "/recommendations?size=20")
                    .check(status().is(200))
                    .check(jsonPath("$.data").exists())
                    .check(bodyString().saveAs("lastRecommendationBody")));

    private final ChainBuilder createWorkOrder =
            exec(http("POST work-order create")
                    .post("/api/v1/work-orders")
                    .body(StringBody("""
                        {
                          "siteId": "cc000000-0000-7206-0001-000000000001",
                          "priority": "HIGH",
                          "description": "Load test synthetic fault #{uuid}",
                          "certificationRequired": "ELEC_DISPATCH"
                        }
                        """.replace("#{uuid}", UUID.randomUUID().toString())))
                    .check(status().in(201, 409)));  // 409 on duplicate reference is acceptable

    private final ChainBuilder searchWorkOrders =
            exec(http("GET work-orders paginated search")
                    .get("/api/v1/work-orders?state=NEW&size=20&page=0")
                    .check(status().is(200))
                    .check(jsonPath("$.page.totalElements").exists()));

    private final ChainBuilder transitionWorkOrder =
            exec(http("PATCH work-order transition to ASSIGNED")
                    .patch("/api/v1/work-orders/" + WO_NEW_2 + "/transitions")
                    .body(StringBody("""
                        {
                          "targetState": "ASSIGNED",
                          "technicianId": "cc000000-0000-7206-0011-000000000001"
                        }
                        """))
                    .check(status().in(200, 409, 422)));  // 409/422 acceptable under concurrent load

    private final ChainBuilder assignWorkOrder =
            exec(http("POST work-order assignment")
                    .post("/api/v1/work-orders/" + WO_NEW_1 + "/assignments")
                    .body(StringBody("""
                        {
                          "technicianId": "cc000000-0000-7206-0011-000000000001"
                        }
                        """))
                    .check(status().in(200, 201, 409, 422)));

    // ── Main scenario (weighted mix) ──────────────────────────────────────────

    private final ScenarioBuilder dispatchMix = scenario("Dispatch recommendation mix")
            .randomSwitch()
            .on(
                50, getRecommendations,
                20, createWorkOrder,
                15, searchWorkOrders,
                10, transitionWorkOrder,
                5,  assignWorkOrder
            );

    // ── Simulation setup ──────────────────────────────────────────────────────

    {
        int warmUpSeconds    = profileInt("warmUpSeconds",    60);
        int rampUpSeconds    = profileInt("rampUpSeconds",    30);
        int holdSeconds      = profileInt("holdSeconds",      120);
        int coolDownSeconds  = profileInt("coolDownSeconds",  15);
        int targetVus        = profileInt("targetVirtualUsers", 40);

        PopulationBuilder population = dispatchMix.injectOpen(
                // Warm-up: ramp to full load, let caches prime — excluded from percentile gate
                rampUsersPerSec(1).to(targetVus).during(Duration.ofSeconds(warmUpSeconds)),
                // Ramp to design load
                rampUsersPerSec(targetVus).to(targetVus * 2).during(Duration.ofSeconds(rampUpSeconds)),
                // Hold at twice peak — this phase produces the gated percentiles
                constantUsersPerSec(targetVus * 2).during(Duration.ofSeconds(holdSeconds)),
                // Cool-down
                rampUsersPerSec(targetVus * 2).to(0).during(Duration.ofSeconds(coolDownSeconds))
        );

        setUp(population)
                .protocols(httpProtocol)
                .assertions(
                        // Phase 2 exit criterion: p95 <= 3 s
                        forAll().responseTime().percentile(95).lt(3001),
                        // Phase 2 exit criterion: server error rate < 1%
                        forAll().failedRequests().percent().lt(1.0)
                );
    }
}

File: app/src/test/load/RecommendationLoadScenario.scala
package load

import io.gatling.core.Predef._
import io.gatling.http.Predef._
import io.gatling.core.structure.ScenarioBuilder
import io.gatling.http.protocol.HttpProtocol

import scala.concurrent.duration._
import scala.io.Source
import com.fasterxml.jackson.databind.ObjectMapper

/**
 * Gatling load scenario for dispatch recommendation latency gate (WO-206).
 *
 * Drives the recommendation endpoint alongside a realistic mix of work-order
 * creation, paginated search, transitions, assignment and parts consumption
 * at twice the design-peak write volume.
 *
 * Configuration is read from:
 *   - src/test/load/config/profiles.json  (ramp, warm-up, hold, cool-down phases)
 *   - src/test/load/config/thresholds.json (p95 ≤ 3s, error rate < 1%)
 *
 * Run locally:
 *   mvn gatling:test -Dgatling.simulationClass=load.RecommendationLoadScenario \
 *       -Dload.profile=healthy \
 *       -Dload.baseUrl=http://localhost:8080 \
 *       -Dload.authToken=<dispatcher-jwt>
 *
 * The warm-up phase is excluded from percentile assertions via Gatling's
 * warmUp / holdFor phases. Metrics are written to target/gatling/results/.
 *
 * Travel provider stub (WireMock) must be running at the URL configured in
 * application-loadtest.yml before starting the scenario.
 *
 * Phase 2 exit criteria evidenced:
 *   - recommendations.p95 ≤ 3000 ms (threshold: recommendations.p95LatencyMs <= 3000)
 *   - server error rate < 1% (threshold: errorRate < 0.01)
 */
class RecommendationLoadScenario extends Simulation {

  // ─── Configuration ────────────────────────────────────────────────────────

  val baseUrl: String  = sys.props.getOrElse("load.baseUrl", "http://localhost:8080")
  val authToken: String = sys.props.getOrElse("load.authToken", "REPLACE_WITH_DISPATCHER_JWT")
  val profileName: String = sys.props.getOrElse("load.profile", "healthy")

  // Load profiles from committed configuration
  val profilesJson: String = Source.fromResource("load/config/profiles.json").mkString

  // Seed work-order IDs and technician IDs from the committed seed SQL
  // (200 technicians seeded by LoadSeedGenerator)
  val workOrderIds: Array[String] = Array(
    "00000000-0000-0000-0001-000000000001",
    "00000000-0000-0000-0001-000000000002",
    "00000000-0000-0000-0001-000000000003",
    "00000000-0000-0000-0001-000000000004",
    "00000000-0000-0000-0001-000000000005"
  )

  val workOrderIdFeeder: Iterator[Map[String, String]] =
    Iterator.continually(workOrderIds.map(id => Map("workOrderId" -> id))).flatten

  val newWorkOrderFeeder: Iterator[Map[String, String]] = Iterator.from(1000).map { i =>
    Map(
      "woTitle"    -> s"Load-test work order $i",
      "woType"     -> "HVAC",
      "priority"   -> "MEDIUM",
      "customerId" -> "00000000-0000-0000-0002-000000000001",
      "siteId"     -> "00000000-0000-0000-0003-000000000001"
    )
  }

  // ─── HTTP protocol ────────────────────────────────────────────────────────

  val httpProtocol: HttpProtocol = http
    .baseUrl(baseUrl)
    .header("Authorization", s"Bearer $authToken")
    .header("Accept", "application/json")
    .header("Content-Type", "application/json")
    .shareConnections
    .connectionHeader("keep-alive")

  // ─── Scenarios ────────────────────────────────────────────────────────────

  val recommendationsScenario: ScenarioBuilder = scenario("Recommendations")
    .feed(workOrderIdFeeder)
    .exec(
      http("GET recommendations")
        .get("/api/v1/work-orders/#{workOrderId}/recommendations")
        .queryParam("pageSize", "20")
        .check(status.in(200, 422)) // 422 when WO not in NEW state is valid product behaviour
        .check(
          jsonPath("$.meta.candidatePoolSize").optional.saveAs("candidatePoolSize")
        )
    )

  val workOrderCreateScenario: ScenarioBuilder = scenario("WorkOrderCreate")
    .feed(newWorkOrderFeeder)
    .exec(
      http("POST work-order")
        .post("/api/v1/work-orders")
        .body(StringBody(
          """{"title":"#{woTitle}","type":"#{woType}","priority":"#{priority}",""" +
          """"customerId":"#{customerId}","siteId":"#{siteId}"}"""
        ))
        .check(status.in(201, 200, 409))
        .check(jsonPath("$.id").optional.saveAs("newWorkOrderId"))
    )

  val workOrderSearchScenario: ScenarioBuilder = scenario("WorkOrderSearch")
    .exec(
      http("GET work-orders (search)")
        .get("/api/v1/work-orders")
        .queryParam("state", "NEW")
        .queryParam("size", "20")
        .queryParam("sort", "createdAt,desc")
        .check(status.is(200))
    )

  val workOrderTransitionScenario: ScenarioBuilder = scenario("WorkOrderTransition")
    .feed(workOrderIdFeeder)
    .exec(
      http("POST transition (START)")
        .post("/api/v1/work-orders/#{workOrderId}/transitions")
        .body(StringBody("""{"event":"START"}"""))
        .check(status.in(200, 409, 422))
    )

  // ─── Phase configuration (read from profiles.json at runtime) ─────────────

  // healthy profile: 20 VUs for 300s hold, 60s ramp, 60s warm-up, 30s cool-down
  // These values mirror profiles.json "healthy" entry
  val rampDuration: FiniteDuration    = 60.seconds
  val warmUpDuration: FiniteDuration  = 60.seconds
  val holdDuration: FiniteDuration    = 300.seconds
  val coolDownDuration: FiniteDuration = 30.seconds
  val targetVirtualUsers: Int         = 20

  // ─── Simulation setup ─────────────────────────────────────────────────────

  setUp(
    // 40% recommendations
    recommendationsScenario.inject(
      nothingFor(rampDuration + warmUpDuration),
      rampUsers(targetVirtualUsers * 4 / 10) during holdDuration
    ),
    // 25% search
    workOrderSearchScenario.inject(
      nothingFor(rampDuration + warmUpDuration),
      rampUsers(targetVirtualUsers * 25 / 100) during holdDuration
    ),
    // 20% create
    workOrderCreateScenario.inject(
      nothingFor(rampDuration + warmUpDuration),
      rampUsers(targetVirtualUsers * 2 / 10) during holdDuration
    ),
    // 10% transitions
    workOrderTransitionScenario.inject(
      nothingFor(rampDuration + warmUpDuration),
      rampUsers(targetVirtualUsers * 1 / 10) during holdDuration
    ),
    // warm-up for all
    recommendationsScenario.inject(
      rampUsers(5) during rampDuration,
      constantUsersPerSec(2) during warmUpDuration
    ).andThen(
      recommendationsScenario.inject(nothingFor(1.second))
    )
  )
    .protocols(httpProtocol)
    .assertions(
      // Phase 2 exit criteria — must mirror thresholds.json
      global.responseTime.percentile(95).lte(3000),
      global.successfulRequests.percent.gte(99.0),
      forAll.responseTime.percentile(95).lte(5000)
    )
}
